### ISSUE TITLE
mm_fp4 trtllm backend leaks padding scales into real rows (use_8x4_sf_layout=True)

### ISSUE THREAD

**Issue by elvircrn** [flashinfer-ai/flashinfer#2861]
### Summary

When `mm_fp4` is called with `backend="trtllm"` and `use_8x4_sf_layout=True` (triggered when `m <= 32`), the kernel reads scale factors from padding rows and mixes them into real rows' output. This causes:

1. **NaN contamination**: If padding scales contain NaN (e.g. from stale `torch.empty` memory under CUDA graphs), real rows' output becomes NaN.
2. **Silent numerical inaccuracy**: Even with finite padding scale values, real rows' output changes (max_diff ~500-700 in bf16).

The CUTLASS backend does not have this issue.

### Reproduction

The test uses swizzle-aware padding detection: quantize the same real rows with two different padding fillers (zeros vs 448.0), diff the resulting scale bytes to find which bytes correspond to padding rows in the swizzled layout, then poison only those bytes.

**NaN contamination test:**

```python
import torch
from vllm._custom_ops import scaled_fp4_quant
from flashinfer import mm_fp4

TILE_ROWS = 128
HIDDEN_DIM = 512
OUT_DIM = 256
BLOCK_SCALE_SIZE = 16

def _round_up(x, multiple):
    return ((x + multiple - 1) // multiple) * multiple

def test(num_real_rows, backend):
    global_scale = torch.tensor([1.0], dtype=torch.float32, device="cuda")
    padded_rows = _round_up(num_real_rows + 1, TILE_ROWS)

    X_pad_zero = torch.zeros(padded_rows, HIDDEN_DIM, dtype=torch.bfloat16, device="cuda")
    X_pad_zero[:num_real_rows] = torch.randn(
        num_real_rows, HIDDEN_DIM, dtype=torch.bfloat16, device="cuda"
    )
    X_pad_448 = X_pad_zero.clone()
    X_pad_448[num_real_rows:] = 448.0

    fp4_data, scales_zero = scaled_fp4_quant(
        X_pad_zero, global_scale, is_sf_swizzled_layout=True, backend=backend
    )
    _, scales_448 = scaled_fp4_quant(
        X_pad_448, global_scale, is_sf_swizzled_layout=True, backend=backend
    )

    padding_mask = (
        scales_zero.view(torch.uint8) != scales_448.view(torch.uint8)
    )

    scales_poisoned = scales_zero.view(torch.uint8).clone()
    scales_poisoned[padding_mask] = 0x7F  # NaN in float8_e4m3fn

    weight_fp4 = torch.randint(
        0, 255, (OUT_DIM, HIDDEN_DIM // 2), dtype=torch.uint8, device="cuda"
    )
    weight_scales = torch.ones(
        OUT_DIM, HIDDEN_DIM // BLOCK_SCALE_SIZE,
        dtype=torch.float8_e4m3fn, device="cuda",
    )
    alpha = torch.tensor([1.0], dtype=torch.float32, device="cuda")
    use_8x4 = (backend == "trtllm" and num_real_rows <= 32)

    output = mm_fp4(
        fp4_data,
        weight_fp4.t(),
        scales_poisoned.view(torch.float8_e4m3fn),
        weight_scales.t(),
        alpha,
        torch.bfloat16,
        block_size=BLOCK_SCALE_SIZE,
        use_8x4_sf_layout=use_8x4,
        backend=backend,
    )

    nan_rows = torch.isnan(output[:num_real_rows]).any(dim=-1).sum().item()
    tag = "FAIL" if nan_rows else "OK"
    print(f"{tag}  m={num_real_rows}  backend={backend}  nan_rows={nan_rows}/{num_real_rows}")

for m in [1, 2, 4, 8, 16, 32, 64]:
    for backend in ["cutlass", "trtllm"]:
        test(m, backend)
```

### Test results

**NaN contamination (padding scales = NaN):**
```
OK    m=  1  backend=cutlass  nan_rows=0/1
FAIL  m=  1  backend=trtllm   nan_rows=1/1
OK    m=  2  backend=cutlass  nan_rows=0/2
FAIL  m=  2  backend=trtllm   nan_rows=2/2
OK    m=  4  backend=cutlass  nan_rows=0/4
FAIL  m=  4  backend=trtllm   nan_rows=4/4
OK    m=  8  backend=cutlass  nan_rows=0/8
FAIL  m=  8  backend=trtllm   nan_rows=8/8
OK    m= 16  backend=cutlass  nan_rows=0/16
FAIL  m= 16  backend=trtllm   nan_rows=14/16
OK    m= 32  backend=cutlass  nan_rows=0/32
FAIL  m= 32  backend=trtllm   nan_rows=24/32
OK    m= 64  backend=cutlass  nan_rows=0/64
OK    m= 64  backend=trtllm   nan_rows=0/64
```

**Finite contamination (padding scales = 1.0 vs original):**
```
SAME  m=  1  backend=cutlass  max_diff=0.000000
DIFF  m=  1  backend=trtllm   max_diff=514.500000
SAME  m=  2  backend=cutlass  max_diff=0.000000
DIFF  m=  2  backend=trtllm   max_diff=692.687500
...
SAME  m= 64  backend=cutlass  max_diff=0.000000
SAME  m= 64  backend=trtllm   max_diff=0.000000
```

### Observations

- **CUTLASS backend**: no cross-row contamination
- **TRT-LLM, m ≤ 32 (use_8x4_sf_layout=True)**: padding scales leak into real rows
- **TRT-LLM, m > 32 (use_8x4_sf_layout=False)**: seems bug is specific to 8×4 SF layout path
- The bug persists even when scales are produced by the correct `flashinfer_quant_nvfp4_8x4_sf_layout` quantization function (passing `backend="trtllm"` to `scaled_fp4_quant`)

### Workaround

Setting padding scales to **zero** (0x00 in float8_e4m3fn) neutralizes the bug: even though the kernel still reads the wrong scales, `0 * data = 0` contributes nothing to real rows' output. This is the same principle behind the `torch.zeros` fix in `create_fp4_scale_tensor` for the CUTLASS MoE path.

### Environment

- GB200 (SM 100, Blackwell)
- vLLM main branch

**Comment by XFDG** [comment]
Hi, I retested the cross-row scale-contamination invariant on an NVIDIA B200 (SM100) with current `main` at `c05407ceffb7d9ce111a73553a8e37ad752adb62`, CUDA 13.1, and TRTLLM-Gen GEMM artifact `7b1fc253cd6237950e76310873f4acf4d97a3904/gemm-b738138-25754e6`.

I kept the first 32 rows' FP4 data and all weights byte-identical, changed only the 8x4-layout scale bytes attributable to rows 32..127, and compared all first-32-row outputs:

```
scale_changed_bytes 3072
real_max_diff 0.0
real_changed 0 / 8192
PASS: no cross-row scale contamination
```

I could not reproduce the reported leak with the current artifact. This appears to have been fixed by a subsequent TRTLLM-Gen GEMM artifact update, so I do not have a justified source patch to submit. If it still reproduces in vLLM, please share the exact FlashInfer commit plus GEMM artifact pin so the older binary can be tested directly.
